### PR TITLE
[FLINK-36284][format-avro] Support extracting Avro type with new DataType stack

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

*Support extracting Avro type in `DataTypeExtractor` to let avro pojo can be used in new DataType stack.*

## Brief change log

  - *Support extracting Avro in `DataTypeExtractor`(refers the logic in TypeExtractor#privateGetForClass).*
  - *Update `AvroTypesITCase` to let it use the new recommended table apis.*

## Verifying this change

Tests are updated to verity this pr.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 1]
```diff
@@ -1971,7 +1967,7 @@ private <OUT, IN1, IN2> TypeInformation<OUT> privateGetForClass(
         }
 
         // special case for POJOs generated by Avro.
-        if (hasSuperclass(clazz, AVRO_SPECIFIC_RECORD_BASE_CLASS)) {
+        if (AvroUtils.isAvroSpecificRecord(clazz)) {
```
Comment (davidradl) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:1970]: I am wondering does does the avro-maven-plugin ever generate / support a POJO that inherits from a POJO class that implements  AVRO_SPECIFIC_RECORD_BASE_CLASS. If so should be walk the inheritance tree to find AVRO_SPECIFIC_RECORD_BASE_CLASS?

Comment (xuyangzhong) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:1970]: Maybe you're right. I've noticed that the generated User class is not marked as final. One case I can think of is creating a non-generated class B that directly extends the generated class A. However, this is not within the scope of the changes in this PR. If you agree, I can create a new Jira ticket to track it.

**Comment by xuyangzhong** [comment]
@davidradl Hi, could you please take another look at my comments on your proposal? This work blocks the removal of the old `toAppendStream` method, which needs to be completed in the 2.0-preview version (deadline is September 30)

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java (hunk 1 above)
Comment (davidradl) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:1970]: @xuyangzhong raising a new Jira to work this separately makes sense to me.

**Review by davidradl — APPROVED** [review]

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java (hunk 1 above)
Comment (xuyangzhong) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:1970]: Trace it at https://issues.apache.org/jira/browse/FLINK-36337

**Review by lincoln-lil — APPROVED** [review]
@xuyangzhong LGTM!

**Pull request merged**